### PR TITLE
Friendly exceptions

### PR DESCRIPTION
**Pull request by viktorprogger**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #25

### CODE REVIEW

**Comment by viktorprogger** [comment]
There is a question: shall I make this exception friendly or remove it? It is from `yii2` version and doesn't thrown anywhere for now. But there is a chance it will be thrown from drivers when serializing/deserializing.

**Comment by viktorprogger** [comment]
[Queue](https://trello.com/c/QTPMAx67/6-queue)

**Comment by samdark** [comment]
> There is a question: shall I make this exception friendly or remove it? It is from yii2 version and doesn't thrown anywhere for now. But there is a chance it will be thrown from drivers when serializing/deserializing.

Where was it used in Yii 2? Is there going to be such situation? If the answer is "no" or "not likely" or even "no idea" then remove it.

File: src/Enum/JobStatus.php [hunk 1]
```diff
@@ -17,7 +17,7 @@ class JobStatus
     protected function __construct(int $status)
     {
         if (!in_array($status, $this->available(), true)) {
-            throw new InvalidArgumentException('Invalid status provided');
+            throw new InvalidStatusException();
```
Comment (samdark) [inline src/Enum/JobStatus.php]: I think the status itself would be very helpful to have in both exception message and exception object.

File: src/Exceptions/InvalidStatusException.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        return 'Invalid status provided';
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getSolution(): ?string
+    {
+        return 'Refer to \Yiisoft\Yii\Queue\Enum\JobStatus::available() to see available status list.';
```
Comment (samdark) [inline src/Exceptions/InvalidStatusException.php]: Can't we list statuses right away so there's no need to open docs?

File: src/Exceptions/JobNotSupportedException.php [hunk 3]
```diff
@@ -18,4 +19,30 @@ public function __construct(DriverInterface $driver, JobInterface $job, int $cod
...
         parent::__construct($message, $code, $previous);
     }
+
+
+    /**
+     * @inheritDoc
+     */
+    public function getName(): string
+    {
+        return 'Job is not supported by current queue driver';
```
Comment (samdark) [inline src/Exceptions/JobNotSupportedException.php]: Can we replace "current queue driver" with actual driver name? Can we replace "Job" with something more concrete?

File: src/Exceptions/JobNotSupportedException.php [hunk 4]
```diff
@@ -18,4 +19,30 @@ public function __construct(DriverInterface $driver, JobInterface $job, int $cod
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            Try to use another driver which supports the given job.
+            Give your attention to the interfaces this job implements.
```
Comment (samdark) [inline src/Exceptions/JobNotSupportedException.php]:
```suggestion
            Pay attention to the interfaces this job implements.
```

File: src/Exceptions/JobNotSupportedException.php (hunk 3 above)
Comment (viktorprogger) [reply src/Exceptions/JobNotSupportedException.php]: Yes, will describe this in `getSolution`.

**Comment by viktorprogger** [comment]
> Where was it used in Yii 2? Is there going to be such situation? If the answer is "no" or "not likely" or even "no idea" then remove it.

Smthng like it will be thrown in drivers, but I can't imagine the concrete usage. Removed for a while.

File: src/Exceptions/JobNotSupportedException.php [hunk 5]
```diff
@@ -6,16 +6,76 @@
...
+
         parent::__construct($message, $code, $previous);
+
+        $this->driver = $driver;
+        $this->job = $job;
+    }
+
+
+    /**
+     * @inheritDoc
```
Comment (samdark) [inline src/Exceptions/JobNotSupportedException.php]: There's no need for these phpdoc blocks.

**Pull request merged**